internal/tools/changes/modules.go

```diff
@@ -34,5 +46,158 @@ func getModFile(path string) (*modfile.File, error) {
+	if modulePath == sdkRepo {
+		return rootModule
+	}
+
+	modulePath = strings.TrimLeft(modulePath, "/")
+	return strings.TrimPrefix(modulePath, sdkRepo+"/")
+}
+
+func lengthenModPath(modulePath string) string {
+	if modulePath == rootModule {
+		return sdkRepo
+	}
+
+	return sdkRepo + "/" + modulePath
+}
+
+// discoverModules returns a list of all modules and a map between all packages and their providing module.
+func discoverModules(golist golist.ModuleClient, root string) ([]string, map[string]string, error) {
+	var modules []string
+	packages := map[string]string{}
+
+	err := filepath.Walk(root, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+
+		if info.Name() == "go.mod" {
+			modFile, err := getModFile(path)
+			if err != nil {
+				return err
+			}
+
+			mod := shortenModPath(modFile.Module.Mod.String())
+
+			modPackages, err := golist.Packages(mod)
+			if err != nil {
+				return err
+			}
+
+			for _, p := range modPackages {
+				packages[p] = mod
+			}
+
+			modules = append(modules, mod)
+		} else if info.Name() == "testdata" && path != "testdata" {
+			// skip testdata directory unless it is the root directory.
+			return filepath.SkipDir
+		}
+
+		return nil
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return modules, packages, nil
+}
+
+func updateDependencies(repoPath, mod string, dependencies []string, enc *VersionEnclosure) error {
+	goModPath := filepath.Join(repoPath, mod, "go.mod")
+
+	modFile, err := getModFile(goModPath)
+	if err != nil {
+		return fmt.Errorf("couldn't update %s's dependencies: %v", mod, err)
+	}
+
+	for _, dep := range dependencies {
+		err = modFile.AddRequire(lengthenModPath(dep), enc.ModuleVersions[dep].Version)
+		if err != nil {
+			return fmt.Errorf("couldn't update %s's dependency on %s: %v", mod, dep, err)
+		}
+	}
+
+	out, err := modFile.Format()
+	if err != nil {
+		return fmt.Errorf("couldn't update %s's dependencies: %v", mod, err)
+	}
+
+	return util.WriteFile(out, goModPath, false)
+}
+
+// defaultVersion returns a default version for the given module based on its import path. If a version suffix /vX is
+// present in the import path, the default version will be vX.0.0. Otherwise, the version will be v0.0.0.
+func defaultVersion(mod string) (string, error) {
+	_, major, ok := module.SplitPathVersion(mod)
+	if !ok {
+		return "", fmt.Errorf("couldn't split module path: %s", mod)
+	}
+	major = strings.TrimLeft(major, "/")
+
+	if major == "" {
+		major = "v0"
+	}
+
+	return fmt.Sprintf("%s.0.0", major), nil
+}
+
+func pseudoVersion(git git.VcsClient, mod string) (string, error) {
+	commitHash, err := git.CommitHash()
+	if err != nil {
+		return "", fmt.Errorf("couldn't make pseudo-version: %v", err)
+	}
+
+	tagVer, err := taggedVersion(git, mod, true)
+	if err != nil {
+		return "", fmt.Errorf("couldn't make pseudo-version: %v", err)
+	}
+
+	return formatPseudoVersion(commitHash, tagVer)
+}
```

Comment:
I think it makes sense to keep it, since it is probably useful to be able to generate service clients depending on commit hashes. We just don't update all the dependencies to be on the latest commit hash. I think the generated service clients would still work, though, and depend on the passed in commit hashes.